source/client-side-operations-timeout/tests/pending-response.yml

```diff
@@ -0,0 +1,349 @@
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *failPointClient
+          failPoint:
+            configureFailPoint: failCommand
+            mode: {times: 1}
+            data:
+              failCommands: ["insert"]
+              blockConnection: true
+              blockTimeMS: 500
+      # Execute operation with timeout less than block time.
+      - name: insertOne
+        object: *collection
+        arguments:
+          timeoutMS: 50
+          document: {_id: 3, x: 1}
+        expectError:
+          isTimeoutError: true
+      # Execute a subsequent operation with a timeout less than the block time.
+      - name: find
+        object: *collection
+        arguments:
+          timeoutMS: 50
+          filter: {_id: 1}
+        expectError:
+          isTimeoutError: true
+      # Execute a final operation to drain the buffer.
+      - name: find
+        object: *collection
+        arguments:
+          filter: {_id: 1}
+    expectEvents:
+      - client: *client
+        events:
+          - commandFailedEvent:
+              commandName: insert
+          - commandSucceededEvent:
+              commandName: find
+      - client: *client
+        eventType: cmap
+        events:
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {} # Insert fails.
+          - connectionPendingResponseStarted: {} # First find fails
+          - connectionPendingResponseFailed:
+              reason: timeout
+          - connectionPendingResponseStarted: {} # Second find drains the buffer.
+          - connectionPendingResponseSucceeded: {}
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {} # Second find succeeds.
+    # If the connection is closed server-side while draining the response, the
+    # driver must retry with a different connection.
+  - description: "read op retries when connection closes server-side while draining response"
+    operations:
+      # Create a failpoint to block the first and second operation.
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *failPointClient
+          failPoint:
+            configureFailPoint: failCommand
+            mode: {times: 1}
+            data:
+              failCommands: ["insert"]
+              blockConnection: true
+              blockTimeMS: 500
+              closeConnection: true
+      # Execute operation with timeout less than block time.
+      - name: insertOne
+        object: *collection
+        arguments:
+          timeoutMS: 50
+          document: {_id: 3, x: 1}
+        expectError:
+          isTimeoutError: true
+      - name: wait
+        object: testRunner
+        arguments:
+          ms: 500
+      # Execute a subsequent operation with a timeout less than the block time.
+      - name: find
+        object: *collection
+        arguments:
+          timeoutMS: 50
+          filter: {_id: 1}
```

Comment:
We've seen flakiness with establishing connections and this test will require a new connection to be established.  Should we up the timeout on the `find` just in case?

Reply:
Maybe I'm confused by how `closeConnection: true` fits into the picture here.  I imagined that:

- failCommand is triggered, write blocks for 500ms 
- client-side, insert times out
- failCommand expires, and connection is closed
- the second insert is executed, checks the connection in the pool, attempts to read the bytes from the socket only to find that the socket has been closed.  the connection is discarded, the checkout request immediately then attempts to open a new connection, which succeeds and the second insert completes successfully

Am I misunderstanding how this test works?

Reply:
Oh, yes! You are right! I've overlooked `closeConnection: true`. Thank you for spotting the problem!